lib/workflow/stores/mongo.js

```diff
@@ -643,18 +643,36 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             graphId: data.graphId,
             reachable: true
         };
-        query['dependencies.' + data.taskId] = {
-            $in: [data.state, Constants.Task.States.Finished]
-        };
         var update = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
         var options = {
             multi: true
         };
+        
+        query['dependencies.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
 
-        return waterline.taskdependencies.updateMongo(query, update, options);
+
+        return waterline.taskdependencies.find(query).then(function(res) {
```

Comment:
You could copy the `updateUnreachableTasks` to use `$or` to connect the two criteria, so that to save this query operation.